scripts/module/formapplications/sequencer-database-viewer.js

```diff
@@ -1,179 +1,228 @@
-            return {
-                pack: entry.split('.')[0],
-                entry: entry
-            };
-        });
-        this.list = false;
-    }
-
-    static show(inFocus = false){
-        if (!game.user.isTrusted) return;
-        for(let app of Object.values(ui.windows)){
-            if(app instanceof this){
-                return app.render(true, { focus: inFocus });
-            }
-        }
-        return new this().render(true);
-    }
+import { cache } from "../lib/cache.js";
+import { reactiveEl as html } from "../lib/html.js";
+import { clamp, strToSearchRegexStr } from "../lib/lib.js";
 
-    static get isVisible(){
-        for(let app of Object.values(ui.windows)){
-            if(app instanceof this) return app;
-        }
-    }
+const MAX_NODES = 24;
 
-    /** @override */
-    static get defaultOptions() {
-        return foundry.utils.mergeObject(super.defaultOptions, {
-            title: "Sequencer Database Viewer",
-            template: `modules/sequencer/templates/sequencer-database-template.html`,
-            classes: ["dialog"],
-            width: 900,
-            height: 425,
+export default class SequencerDatabaseViewer extends FormApplication {
+  constructor(dialogData = {}, options = {}) {
+    super(dialogData, options);
+    this.filter = "all";
+    this.search = "";
+    this.autoplay = true;
+    this.packs = Object.keys(window.Sequencer.Database.entries);
+    this.entries = window.Sequencer.Database.flattenedEntries.map((entry) => {
+      return {
+        pack: entry.split(".")[0],
+        entry: entry,
+      };
+    });
+
+    // cache getFilteredEntries method, breaking cache whenever search or filter property changes
+    cache(this, "getFilteredEntries", ["search", "filter"]);
+    cache(this, "getSearchRegex", ["search"]);
+
+    this.listItems = Array.from(
+      { length: Math.min(this.getFilteredEntries().length, MAX_NODES) },
+      () => {
+        const el = html`<div class="database-entry" data-id="{{ entry }}">
+          <button type="button" class="btn_play">
+            <i class="fas fa-play"></i>
+          </button>
+          <button type="button" class="btn_copy_filepath">
+            <i class="fas fa-copy"></i> Filepath
+          </button>
+          <button type="button" class="btn_copy_databasepath">
+            <i class="fas fa-copy"></i> Database
+          </button>
+          <div class="database-entry-text" title="{{entry}}">
+            <div class="database-entry-text-highlight"></div>
+            {{ entry }}
+          </div>
+        </div>`;
+
+        const highlight = el.querySelector(".database-entry-text-highlight");
+
+        el.addEventListener("update", (e) => {
+          highlight.replaceChildren();
+          if (!this.search) return;
+          const entry = e.detail.data?.entry ?? "";
+          const regex = this.getSearchRegex();
+
+          highlight.innerHTML = entry.replace(regex, "<mark>$&</mark>");

```

Comment:
Guide: This is really to fix a problem created by the fact that I'm editting the text in hte HTML directly so we don't want to remove the node and mess with them. Instead the highlighting actually occurs on an element thats ontop with invisible text that receives the highlighting instead. seems to work pretty well